File: Documenti/Git Repos/MainRepo/Potential_Field/src/potential_field.cpp
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/PoseStamped.h>
#include <nav_msgs/Odometry.h>
#include <geometry_msgs/Twist.h>
#include <sensor_msgs/LaserScan.h>
#include <visualization_msgs/MarkerArray.h>
#include <visualization_msgs/Marker.h>
#include <tf/transform_datatypes.h>
#include <ros/ros.h>
#include <vector>

#include "potential_field.hpp"


void PotentialField::initializePublishers() 
{
    velocityPublisher = nh.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
    markerArrayPublisher = nh.advertise<visualization_msgs::MarkerArray>("/visualization_marker_array", 1);
}

void PotentialField::initializeSubscribers() 
{
    odomSubscriber = nh.subscribe<nav_msgs::Odometry>("/odom", 1, this);
    laserSubscriber = nh.subscribe<sensor_msgs::LaserScan>("/kobuki/laser/scan", 1, this);
    goalSubscriber = nh.subscribe<geometry_msgs::PoseStamped>("/visualization_marker_array", 1, this);
}

void PotentialField::goalCB(const geometry_msgs::PoseStamped::ConstPtr& msg) 
{
    goalX = msg->pose.position.x;
    goalY = msg->pose.position.y;
    goalPublished = true;
    ROS_INFO("Goal has been published => x: %f, y: %f", goalX, goalY);
}

void PotentialField::laserCB(const sensor_msgs::LaserScan::ConstPtr& msg)
{
    laserReadings = msg->ranges;
    ROS_INFO("Laser Range Finder has started publishing...");
}

void PotentialField::odomCB(const nav_msgs::Odometry::ConstPtr& msg)
{
    //update x, y and yaw robot odometry values 
    x = msg->pose.pose.position.x;
    y = msg->pose.pose.position.y;
    tf::Quaternion q(msg->pose.pose.orientation.x, msg->pose.pose.orientation.y, 
        msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);
    tf::Matrix3x3 m(q);
    double roll, pitch;
    m.getRPY(roll, pitch, theta);
   
   //check if odometry has started publishing and print appropriate messages 
    if (odomCounter == 0)
        ROS_INFO("Odometry has started publishing...");
        ROS_INFO("Initial robot pose => x: %f, y: %f, yaw: %f", x, y, theta);
     ++odomCounter;
}

void PotentialField::publishOnce(geometry_msgs::Twist vel)
{
    ros::Rate loop_rate(20);
    while (ros::ok()) 
    {
        int connections = velocityPublisher.getNumSubscribers();
        if (connections > 0) {
            velocityPublisher.publish(vel);
            break;
        }
        loop_rate.sleep();
    }
}

std::vector<double> PotentialField::potentialField(double xi, double eta, double distanceOfInfluence)
{
    double attractivePotentialX = xi * (x - goalX);
    double attractivePotentialY =  xi * (y - goalY);
    double repulsivePotentialX(0.0);
    double repulsivePotentialY(0.0);

    for (std::vector<float>::const_iterator it = laserReadings.begin() ; it != laserReadings.end(); ++it) {
        if (*it <= 30.0) {
            double obstacleAngle = ((it - laserReadings.begin()) - 360) * M_PI / 720;
            double distanceToObstacleX = x + *it * cos(theta + obstacleAngle);
            double distanceToObstacleY = y + *it * sin(theta + obstacleAngle);

            repulsivePotentialX += eta * (1 / *it - 1 / distanceOfInfluence) * (1 / std::pow(*it, 3)) * (x - distanceToObstacleX);
            repulsivePotentialY += eta * (1 / *it - 1 / distanceOfInfluence) * (1 / std::pow(*it, 3)) * (y - distanceToObstacleY);
        }
    }

    double potentialFieldX = attractivePotentialX + repulsivePotentialX;
    double potentialFieldY = attractivePotentialY + repulsivePotentialY;
    double magnitude = std::sqrt(std::pow(potentialFieldX, 2) + std::pow(potentialFieldY, 2));
    double direction = std::atan2(potentialFieldX, potentialFieldY);

    std::vector<double> result = {potentialFieldX, potentialFieldY, magnitude, direction};
    return result;  
}

double PotentialField::euclideanDistance(double targetX, double targetY) 
{
    return std::sqrt(std::pow(targetX - x, 2) + std::pow(targetY -y, 2));
}

double PotentialField::angularVelocity(double targetTheta, double constant = 2.0)
{
    double angularVelocity = std::abs(targetTheta - theta);
    if (theta > targetTheta)
        angularVelocity = -1 * angularVelocity;
    return constant * angularVelocity;
}

void PotentialField::PIDcontroller(double targetX, double targetY, double targetTheta, double distanceTolerance)
{
    geometry_msgs::Twist vel;
    ros::Rate loop_rate(20);

    visualization_msgs::Marker stepGoalMarker;
    stepGoalMarker.type = visualization_msgs::Marker::CUBE;
    stepGoalMarker.action = visualization_msgs::Marker::ADD;
    stepGoalMarker.scale.x = 0.2;
    stepGoalMarker.scale.y = 0.2;
    stepGoalMarker.scale.z = 0.2;
    stepGoalMarker.header.frame_id = "odom";
    stepGoalMarker.color.r = 1.0;
    stepGoalMarker.color.g = 0.0;
    stepGoalMarker.color.b = 0.0;
    stepGoalMarker.color.a = 0.8;
    stepGoalMarker.pose.position.x = x;
    stepGoalMarker.pose.position.y = y;

    visualization_msgs::Marker vectorMarker;
    stepGoalMarker.type = visualization_msgs::Marker::ARROW;
    stepGoalMarker.action = visualization_msgs::Marker::ADD;
    stepGoalMarker.scale.x = 0.75;
    stepGoalMarker.scale.y = 0.2;
    stepGoalMarker.scale.z = 0.2;
    stepGoalMarker.header.frame_id = "odom";
    stepGoalMarker.color.r = 0.0;
    stepGoalMarker.color.g = 1.0;
    stepGoalMarker.color.b = 0.0;
    stepGoalMarker.color.a = 0.8;
    stepGoalMarker.pose = odomPose;
    stepGoalMarker.pose.orientation.z = theta;

    visualization_msgs::Marker finalGoalMarker;
    stepGoalMarker.type = visualization_msgs::Marker::SPHERE;
    stepGoalMarker.action = visualization_msgs::Marker::ADD;
    stepGoalMarker.scale.x = 0.2;
    stepGoalMarker.scale.y = 0.2;
    stepGoalMarker.scale.z = 0.2;
    stepGoalMarker.header.frame_id = "odom";
    stepGoalMarker.color.r = 0.0;
    stepGoalMarker.color.g = 0.0;
    stepGoalMarker.color.b = 1.0;
    stepGoalMarker.color.a = 0.8;
    stepGoalMarker.pose.position.x = goalX;
    stepGoalMarker.pose.position.y = goalY;

    visualization_msgs::MarkerArray markerArray;
    markerArray.markers.push_back(stepGoalMarker);
    markerArray.markers.push_back(vectorMarker);
    markerArray.markers.push_back(finalGoalMarker);

    int id = 0;
    for (int i = 0; i != markerArray.markers.size; ++i) {
        markerArray.markers.at(i).id = id;
        ++i;
    }

    int count = 0;
    double initialDistance = euclideanDistance(targetX, targetY);
    while ((euclideanDistance(targetX, targetY) >= distanceTolerance) && 
        (euclideanDistance(targetX, targetY) <= initialDistance))
    {
        distancesToGoal.push_back(euclideanDistance(goalX, goalY));
        markerArrayPublisher.publish(markerArray);

        vel.linear.x = 0.3;
        vel.linear.y = 0.0;
        vel.linear.z = 0.0;

        vel.angular.x = 0.0;
        vel.angular.y = 0.0;
        vel.angular.z = angularVelocity(targetTheta);

        velocityPublisher.publish(vel);

        ++count;

        if (distancesToGoal.size() > 20) {
            double distanceTravelled = std::abs(distancesToGoal.at(distancesToGoal.size() - 1)
                - distancesToGoal.at(distancesToGoal.size() - 21));
            if ((distanceTravelled < 0.05) && (euclideanDistance(goalX, goalY) > 1.0)) {
                escapeLocalMinima();
            }
        loop_rate.sleep();
        }
    }
}

void PotentialField::gradientDescentWithPID(double step=0.0075, double xi=20.0, 
    double eta=0.9, double distanceOfInfluence=75.0)
{
    geometry_msgs::Twist vel;
    int i = 0;
    double stepGoalX;
    double stepGoalY;

    while (ros::ok()) 
    {
        if (goalPublished == true) {
            std::vector<double> potentialFieldVector(potentialField(xi, eta, distanceOfInfluence));
            odomPose.orientation.z = potentialFieldVector.at(3);

            double stepGoalTheta = 0.0;

            if (i == 0) {
                stepGoalX = x;
                stepGoalY = y;
            }

            double distanceToGoal = std::sqrt(std::pow(goalX - stepGoalX, 2) + std::pow(goalY - stepGoalY, 2));

            if (distanceToGoal > 0.25) {
                stepGoalX += step * (potentialFieldVector.at(0) / potentialFieldVector.at(2));
                stepGoalY += step * (potentialFieldVector.at(1) / potentialFieldVector.at(2));
                stepGoalTheta = potentialFieldVector.at(3);
            } else {
                goalPublished = false;
                continue;
            }
            ++i;

            PIDcontroller(stepGoalX, stepGoalY, stepGoalTheta, 0.175);

            if (euclideanDistance(stepGoalX, stepGoalY) > 0.4) 
                i = 0;

        }  else {
            vel.linear.x = 0.0;
            vel.angular.z = 0.0;
            velocityPublisher.publish(vel);
            distancesToGoal.clear();
        }
    }
    ros::spin();
}

void PotentialField::escapeLocalMinima() 
{
    geometry_msgs::Twist vel;
    while (laserReadings.at(laserReadings.size() / 2) <= 30.0) 
    {
        vel.angular.z = 5.0;
        velocityPublisher.publish(vel);
        distancesToGoal.clear();
    }

    vel.angular.z = 0.0;
    publishOnce(vel);

    ros::Time start = ros::Time::now();
    while (ros::Time::now() < (start + ros::Duration(4))) 
    {
        vel.linear.x = 0.5;
        velocityPublisher.publish(vel);
    }
    vel.angular.z = 0.0;
    publishOnce(vel);
}

int main(int argc, char** argv) 
{
    ros::NodeHandle nodeHandle;
    ros::init(argc, argv, "potential_field_node");
    PotentialField potentialField(nodeHandle);
    potentialField.gradientDescentWithPID(0.0075, 20.0, 0.9, 75.0);
    return 0;
}

File: Documenti/Git Repos/MainRepo/Potential_Field/src/potential_field.py
#! /usr/bin/env python

import rospy
from sensor_msgs.msg import LaserScan
import geometry_msgs
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Twist, Vector3
from std_msgs.msg import Header, ColorRGBA
from visualization_msgs.msg import Marker
from visualization_msgs.msg import MarkerArray
from tf.transformations import euler_from_quaternion
from math import pow, pi, cos, sin, sqrt, atan2, fabs
import numpy as np

class PotentialField():
    def __init__(self):
        #initialize node
        rospy.init_node("obstacle_avoidance_turtlebot")

        #initialize subscribers and publishers
        self.laser_subscriber = rospy.Subscriber("/kobuki/laser/scan", LaserScan, self.laser_cb)
        self.odom_subscriber = rospy.Subscriber("odom", Odometry, self.odom_cb)
        self.publisher = rospy.Publisher("cmd_vel", Twist, queue_size=1)
        self.goal_subscriber = rospy.Subscriber("/potential_field_goal", PoseStamped, self.goal_cb)
        self.marker_array_publisher = rospy.Publisher("/visualization_marker_array", MarkerArray, queue_size=1)

        #initialize robot odometry variables 
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.odom_pose = geometry_msgs.msg.Pose()
        self.odom_counter = 0

        #initialize robot laser variables 
        self.laser_readings = []

        #initialize goal pose variables 
        self.goal_x = 0.0
        self.goal_y = 0.0
        self.goal_pose = geometry_msgs.msg.Pose()
        self.goal_published = False

        #loop rate in Hz
        self.rate = rospy.Rate(20)

        #initialize variable used for escaping local minima 
        self.distances_to_goal = []
        
    #subscriber to topic in which rviz 2D Nav Goal /potential_field_goal publishes into 
    def goal_cb(self, msg):
        self.goal_x = msg.pose.position.x
        self.goal_y = msg.pose.position.y
        self.goal_published = True
        rospy.loginfo('Goal has been published => x: %f, y: %f' %(self.goal_x, self.goal_y))

    #Hokuyo laser range finder subscriber 
    def laser_cb(self, msg):
        self.laser_readings = msg.ranges

    #robot odometry subscriber; in order to not have spammy odometry messages, only the first is printed 
    #to make the code user aware that the robot started publishing the odometry 
    def odom_cb(self, msg):
        self.x = msg.pose.pose.position.x
        self.y  = msg.pose.pose.position.y
        orientation = msg.pose.pose.orientation
        (_, _, self.theta) = euler_from_quaternion([orientation.x, orientation.y,
            orientation.z, orientation.w])
        self.odom_pose = msg.pose.pose
        self.odom_counter += 1
        if self.odom_counter == 1:
            rospy.loginfo('Initial robot pose => x: %f, y: %f, yaw: %f' %(self.x, self.y, self.theta))
            rospy.loginfo('Odometry is being published...')
    
    #the function waits for at least for one subscriber to be active to publish a message before publishing 
    #it is useful when only one publish call is done usually outside a while loop 
    def publish_once(self, rotate):
        while not rospy.is_shutdown():
            connections = self.publisher.get_num_connections()
            if connections > 0:
                self.publisher.publish(rotate)
                break
            self.rate.sleep()
    
    #returns the potential field vector (magnitude, direction and potential field in x and y directions)
    def potential_field(self, xi , eta, distance_of_influence):
        attractive_potential = -xi * np.asarray([self.x - self.goal_x, self.y - self.goal_y])
        repulsive_potential = 0.0
        
        for i, reading in enumerate(self.laser_readings):
            if reading <= 30.0:
                obstacle_angle = (i - 360) * pi / 720
                q_obstacle_x = self.x + reading * cos(self.theta + obstacle_angle)
                q_obstacle_y = self.y + reading * sin(self.theta + obstacle_angle)
                if reading <= distance_of_influence:
                    repulsive_potential += eta * (1 / reading - 1 / distance_of_influence) * (1 / pow(reading, 3)) * np.asarray([self.x - q_obstacle_x, self.y - q_obstacle_y])
                elif reading > distance_of_influence:
                    repulsive_potential += 0

        potential_field = attractive_potential + repulsive_potential
        magnitude = sqrt(pow(potential_field[0], 2) + pow(potential_field[1], 2))
        direction = atan2(potential_field[1], potential_field[0])
        #rospy.loginfo("The potential field in the x direction is %f, while in the y direction is %f" %(potential_field[0], potential_field[1]))
        #rospy.loginfo("The magnitude of the potential field vector is %s while the direction in degrees (0-360) is %s" % (magnitude, (direction + 2*pi) * 180 / pi if direction < 0 else direction * 180 / pi))
        return (potential_field[0], potential_field[1], magnitude, direction)

    #euclidean distance from current position to a certain (x,y) position
    def euclidean_distance(self, x, y):
        return sqrt(pow((x - self.x), 2) + pow((y - self.y), 2))

    #PID-controlled angular velocity
    def angular_velocity(self, theta, constant=2):
        angular_vel = fabs(theta - self.theta)
        if self.theta > theta:
            angular_vel = -angular_vel
        return constant * angular_vel

    #Proportional Integral Differential Controller to get to a certain pose 
    def PID_controller(self, x, y, theta, distance_tolerance):
        vel = Twist()

        #create an rviz marker for the goal position and set the pose 
        step_goal_marker = Marker(type=Marker.CUBE, action=Marker.ADD, scale=Vector3(0.2, 0.2, 0.2),
            header=Header(frame_id='odom'), color=ColorRGBA(1.0, 0.0, 0.0, 0.8))
        step_goal_marker.pose.position.x = x
        step_goal_marker.pose.position.y = y
        #rospy.loginfo("step x is %f and step y is %f" %(x, y))

        #create an rviz marker for the potential field vector and set the direction
        vector_marker = Marker(type=Marker.ARROW, action=Marker.ADD, scale=Vector3(0.75, 0.2, 0.2),
            header=Header(frame_id='odom'), color=ColorRGBA(0.0, 1.0, 0.0, 0.8))
        self.odom_pose.orientation.z = theta
        vector_marker.pose = self.odom_pose

        #create an rviz marker for the final goal position 
        final_goal_marker = Marker(type=Marker.SPHERE, action=Marker.ADD, scale=Vector3(0.2, 0.2, 0.2),
            header=Header(frame_id='odom'), color=ColorRGBA(0.0, 0.0, 1.0, 0.8))
        final_goal_marker.pose.position.x = self.goal_x
        final_goal_marker.pose.position.y = self.goal_y
                
        #create rviz marker array and append the three rviz markers showing potential field vector, step and final goal position
        marker_array = MarkerArray()
        marker_array.markers.append(step_goal_marker)
        marker_array.markers.append(vector_marker)
        marker_array.markers.append(final_goal_marker)

        #number the markers
        id = 0
        for m in marker_array.markers:
            m.id = id
            id +=1 
        
        count = 0
        initial_distance = self.euclidean_distance(x, y)
        while (self.euclidean_distance(x, y) >= distance_tolerance) and (self.euclidean_distance(x, y) <= initial_distance):
            self.distances_to_goal.append(self.euclidean_distance(self.goal_x, self.goal_y))

            #publish the marker array 
            self.marker_array_publisher.publish(marker_array)

            #rospy.loginfo("entering pid controller loop...")
            #rospy.loginfo("next goal in x: %f and y: %f" %(x, y))
            #rospy.loginfo("current x: %f, current y: %f" % (self.x, self.y))
            #rospy.loginfo("distance to next goal point is %f and distance tolerance is %f" % (self.euclidean_distance(x, y), distance_tolerance))
            #linear velocity in x axis
            vel.linear.x = 0.3
            vel.linear.y = 0
            vel.linear.z = 0

            #angular velocitu in z axis
            vel.angular.x = 0
            vel.angular.y = 0
            vel.angular.z = self.angular_velocity(theta)

            #publish velocity command
            self.publisher.publish(vel)

            #increase count
            count += 1

            #analyze the last 10 
            if len(self.distances_to_goal) > 20:
                distance_travelled = fabs(self.distances_to_goal[-1] - self.distances_to_goal[-21])
                if distance_travelled < 0.05 and self.euclidean_distance(self.goal_x, self.goal_y) > 1.0:
                    self.escape_local_minima()

            self.rate.sleep()

    #potential field gradient descent algorithm with PID-controlled speeds
    def gradient_descent_with_PID(self, step=0.0075, xi=20.0, eta=0.9, distance_of_influence=0.85):
        vel = Twist()
        i = 0       
        while not rospy.is_shutdown():
            if self.goal_published == True:
                #generate potential field 
                (f_x, f_y, magnitude, direction) = self.potential_field(xi, eta, distance_of_influence)
                self.odom_pose.orientation.z = direction

                #if it is the first iteration, set the next goal as robot starting position
                if i == 0:
                    q_x = self.x
                    q_y = self.y

                q_theta = 0.0
                distance_to_goal = sqrt(pow((self.goal_x - q_x), 2) + pow((self.goal_y - q_y), 2))

                #rospy.loginfo("distance from current location to final goal is %f" % distance_to_goal)
                if distance_to_goal > 0.25: #add some tolerance
                    q_x += step * (f_x / magnitude)
                    q_y += step * (f_y / magnitude)
                    q_theta = direction
                else:
                    self.goal_published = False
                    continue
                i += 1

                #Proportional Integral Derivative (PID) controller to drive the robot through the
                #generated path plan
                #rospy.loginfo("new step goal point generated...")
                self.PID_controller(q_x, q_y, q_theta, 0.175)

                if self.euclidean_distance(q_x, q_y) > 0.4:
                    i = 0

            else:
                #stop robot once goal is reached                 
                vel.linear.x = 0
                vel.angular.z = 0
                self.publisher.publish(vel)
                self.distances_to_goal = []

        #when Ctrl-C, the node gets killed, else a new goal can be published and the process can re-begin
        rospy.spin()

    #simple function to escape local minima by moving in the direction of no obstacle and then giving back 
    #control to the PID controller
    def escape_local_minima(self):
        #rotate until no obstacle is detected in the middle value of the laser range finder array then move for 4 seconds straight ahead
        vel = Twist()
        while self.laser_readings[len(self.laser_readings) / 2] <= 10.0:
            vel.angular.z = 5.0
            self.publisher.publish(vel)
            self.distances_to_goal = []

        vel.angular.z = 0.0
        self.publish_once(vel)

        start = rospy.Time().now()
        while rospy.Time().now() < (start + rospy.Duration(4)):
            vel.linear.x = 0.5
            self.publisher.publish(vel)
        vel.angular.z = 0.0
        self.publish_once(vel)   
        #once the function has finished executing, the control is given back to the PID and the potential field algorithm 

            
if __name__ == "__main__":
    pf = PotentialField()
    pf.gradient_descent_with_PID(xi=0.15, eta=0.0095, distance_of_influence=4.0)
